### ISSUE TITLE
Error spam (and/or crash) in release template / exported builds: `"slot >= slot_max" is true. Returning: nullptr`

### ISSUE THREAD

**Issue by Dunkhan** [godotengine/godot#70910]
### Godot version

4.0 beta 10

### System information

Windows 10 

### Issue description

In my rss reader on windows, when the user switches between feed lists - that is, when a VBoxContainer has all its children removed and a new set re-added, the application crashes with:

USER ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:974)

This only happens on the windows version and only when that version is built in release mode. In an earlier beta (7 or 10 I am not sure) it was happening occasionally on the linux build too, but that no longer appears to be an issue. I was not able to find any information or other bug reports about this error.

### Steps to reproduce

I have no simple minimal reproduction available, but it is fairly constant in my current project. You can find the source here:
https://gitlab.com/gametheatre/poriferareader/Porifera

1. Build a release windows version of the project
2. Run the exported executable
3. After running the project, subscribe to two or more RSS feeds.
4. If it still has not crashed, switch back and forth between these feeds.

Because this appears to be an issue with the build system it is likely that it will be a major hassle to debug. Let me know if there is any help I can offer.

### Minimal reproduction project

As mentioned above it is consistently reproducible with this (non minimal) project: https://gitlab.com/gametheatre/poriferareader/Porifera

As a temporary workaround, only creating debug builds for windows solves this.

**Comment by Dunkhan** [comment]
In general I feel like a lot of Godot error messages that come from the c++ code could be a bit more verbose. As far as I can tell from the code (I am not great at c++) it has something to do with signals.

**Comment by Dunkhan** [comment]
Looking into it further, the objects being created are subscribing to events. Is it possible that there is an issue with having too many subscriptions to the same event? If so I can work out a different way of passing these changes around.
I should mention I am using C# Mono, hence the word events instead of signals. The event being subscribed to is not a native Godot event and therefore is unlikely to have a signal component. On the other hand I have no idea how mono is integrated so maybe it does.

**Comment by Dunkhan** [comment]
This appears to be fixed in beta 12. I will do some more testing and confirm

**Comment by Dunkhan** [comment]
Still happening on windows, seems to be fixed on linux though

**Comment by akien-mga** [comment]
When you say "built on a Linux system", do you mean exporting the project with the official pre-compiled export templates, or are you compiling custom export templates with mingw?

**Comment by Dunkhan** [comment]
exporting the project with the official pre-compiled export templates

**Comment by akien-mga** [comment]
I tried to reproduce it by exporting from Linux with 4.0 beta 13, and running the project in Wine, but it doesn't seem to crash.
![image](https://user-images.githubusercontent.com/4701338/213155578-25f68cb3-9233-4a20-9b20-97a78f6f8264.png)

As a side note, I had an error when exporting:
```
ERROR: Cannot open file 'res://Themes/MainTheme.tres'.
   at: load (scene/resources/resource_format_text.cpp:1581)
ERROR: Failed loading resource: res://Themes/MainTheme.tres. Make sure resources have been imported by opening the project in the editor at least once.
   at: _load (core/io/resource_loader.cpp:214)
ERROR: res://Item.tscn:10 - Parse Error: [ext_resource] referenced non-loaded resource at: res://Themes/MainTheme.tres
   at: _parse_ext_resource (scene/resources/resource_format_text.cpp:178)
ERROR: res://Item.tscn:47 - Parse Error: [ext_resource] referenced non-loaded resource at: res://Themes/MainTheme.tres
   at: _parse_ext_resource (scene/resources/resource_format_text.cpp:178)
```
But it's not necessarily related. This file is missing your GitLab repo.

Also had this error, which I suspect is a bug in the Mono module:
```
editor/export/editor_export_plugin.h:123 - Required virtual method EditorExportPlugin::_get_name must be overridden before calling.
```

**Comment by Dunkhan** [comment]
thanks for trying, while I was thinking about your last question it occurred to me that I started exporting in a slightly different way at around the same time this started happening. I wrote a bash script that deploys the whole project. The script includes the lines:

```
rm -r $windowsBuildPath
mkdir $windowsBuildPath
$godot --export-release "Windows Desktop" $windowsBuildPath/Porifera.exe
```
Is this correct? Is it possible that this is required for the reproduction?
If there is a possibility of that I can do some tests to check.

I have no idea why it is still asking for 'res://Themes/MainTheme.tres', that was removed some time ago and I can't find any remaining references to it.

**Comment by Dunkhan** [comment]
My first test seems to confirm it is the script. Is it maybe the fact that the script builds release and when I do it by hand the default is debug?

**Comment by akien-mga** [comment]
`--export-release` makes a release build yes, and in the export dialog in the editor there's a checkbox "Export with Debug". If you didn't disable it, then it's a debug build. You should see "(DEBUG)" in the window title.

**Comment by Dunkhan** [comment]
I was more asking if that could be causing the crash. I suppose I need to test that for myself though.

**Comment by akien-mga** [comment]
Yes a release export is more likely to crash than a debug export, as it does less validation checks to save performance. But it would usually point out the error in debug mode.

**Comment by PauloVBettio** [comment]
This is happening to me as well. Running it through Wine does not throws any errors on the console, but running it on an actual Windows machine does, and crashes the game. 

Windows:
![image](https://user-images.githubusercontent.com/74562432/213825041-561c288f-4954-43aa-b38c-d6cc898cf921.png)

Linux / Wine
![image](https://user-images.githubusercontent.com/74562432/213825069-e4bc2d72-f235-4496-a032-10bd2e152157.png)

I've used Linux to export my game with the official beta 14 export template.

**Comment by Dunkhan** [comment]
I did further tests and I am able to build a debug version on linux for windows and it does not crash. I will do a test to see how the windows -> windows release build does.

**Comment by PauloVBettio** [comment]
Building the game on an Windows machine gave me the same result as @Dunkhan. The game only shows the splash screen and crashes.

**Comment by shendo** [comment]
I encountered a similar issue with an Android project on Godot 4 recently.

No noticed problems until switching to release builds which would have lots of screen layout issues and occasional crashes, accompanied by the ‘slot_max’ error logs.

I believe the error message is just a symptom of referencing an invalid object.

In my case I was adding and removing Controls to Containers with GDScript but invalidly using Object.free() on the child before it was implicitly removed.  Presumably, this was causing the container to read the already free’d memory with undefined results.  The debug builds probably have some safety checks which result in the memory staying valid for longer.

Changing my code to explicitly call Container.remove_child() and use child.queue_free() fixed the issue.

**Comment by Dunkhan** [comment]
So you have to remove the child and then free it? Why is there nothing about this in the documentation? [This article](https://godotlearn.com/godot-3-1-how-to-destroy-object-node/) specifically does not mention this. Neither does the [node page](https://docs.godotengine.org/en/stable/classes/class_node.html).

**Comment by h0lley** [comment]
I am also getting spammed with
```
ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:972)
```
on Ubuntu when a level is being loaded with a `template_release` build.
I reckon there's probably no single solution to it but it's project individual.

so the main issue seems to me that this error message is entirely cryptic to users.

**Comment by Dunkhan** [comment]
I noticed I am also getting this error on linux builds, regardless of release or debug. There appear to be no symptoms in the programs behaviour though.

**Comment by YuriSizov** [comment]
> so the main issue seems to me that this error message is entirely cryptic to users.

According to the code comments, "This should never happen unless RID is corrupted." So it cannot really be made less cryptic, since this seems to be a critical failure causing invalid data. Would be amazing to get a small, purpose-built MRP to find the culprit.

**Comment by Dunkhan** [comment]
> Would be amazing to get a small, purpose-built MRP to find the culprit.

The stability of the last few beta releases with my project, and the amount of things that have randomly broken with no documentation or debug output, suggest to me that we should probably be testing with larger projects like this more. A small purpose built MRP is great for resolving a single issue, if the problem can be narrowed down that far, but there is no substitute for testing with scenarios that mirror real world usage. I predict that a lot of hidden bugs that only manifest in more complex use-cases are going to make the 4.0 release a real headache for a lot of people that migrate on release day.

**Comment by Brawmario** [comment]
I'd like to chime in, I'm also having this exact error when running the demo for [my libsm64-godot addon](https://github.com/Brawmario/libsm64-godot) on Linux every time, both in debug and release, but never on my Windows build. The addon has a GDExtension module in it which is compiled through the included GitHub action in the .github folder on a Ubuntu runner.

**Comment by sine707** [comment]
I also had this error occur when building for Windows Desktop and Web, building from Windows 10 on Godot 4.0 stable.

It happened when I cleared objects in a UI container with a for loop.
I eventually managed to get rid of it by replacing `free()` with `queue_free()`

Debug builds ran fine without errors, only release builds seem to be affected which makes it a particularly nasty bug.

It's easy to reproduce, here's a quick example.
A few iterations seem fine, but a large amount seems to spark the error.
```
var number_of_iterations: int = 300

func clear_and_repopulate() -> void:
	if container.get_child_count() > 0:
		for square in container.get_children():
			square.free()
	
	populate_squares()


func populate_squares() -> void:
	for _i in number_of_iterations:
		var square: ColorRect = ColorRect.new()
		container.add_child(square)
```

**Comment by SpaceAttorney** [comment]
Getting this error too. Windows 10, Godot 4.0.stable. Only have debug builds so far and don't get the error every time. Happens sometimes when I try to initiate dialogue. Error message prints twice per crash (to console only, not debugger), and very rarely not at all. I'm using a plugin for dialogue and so don't have a complete handle on the code, but as far as I can trace it, nothing's getting deleted that could be causing a null reference.

**Comment by lorenzo-arena** [comment]
I'm getting this error in 4.0.1 stable (using the .NET version), and only in release exports of my game. One of my scenes has some placeholder objects in a `GridContainer` which are freed in `_Ready`; switching to that scene than to another and back again for a few times seems to trigger the error. I'm getting nothing printed to the console but by enabling file logs in the last `godot.log` I found the same:

```
USER ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:972)
```

after that, the games hangs for a couple seconds then crashes.

**Comment by lorenzo-arena** [comment]
After removing the placeholder objects from the scene I'm still getting crashes with

```
USER ERROR: Condition "!canvas_item" is true.
   at: canvas_item_clear (servers/rendering/renderer_canvas_cull.cpp:1642)
```

Though I don't know if it's related at this point

**Comment by terminator98202** [comment]
I'm having similar issues https://github.com/godotengine/godot/issues/75802

**Comment by Dunkhan** [comment]
So I found an issue where there is an exception on Linux as well in the release version that is not present in the debug version. I also noticed that the GD.Print logger crashes when it happens and ceases mid line to output anything. It is not clear that the slot >= slot_max errors actually have anything to do with the symptoms any of us are experiencing, I wonder if I should rename the issue...

For the moment I am now releasing debug versions on both platforms.

**Comment by SpaceAttorney** [comment]
Still getting this error, but I figured out something strange: when I add breakpoints or print statements to the function I suspect the crash occurs in, the crash _stops occurring_. This is intensely aggravating from a troubleshooting standpoint, but interesting enough to be worth reporting in its own right.

**Comment by dmitriy-shmilo** [comment]
Reproduces with Godot 4.0.3 (Steam). Exporting for Windows, release. Both when exporting on a Windows and a Mac OS machine. The full log when running my exported application and performing steps to reproduce is just this:

```
Godot Engine v4.0.3.stable.official.5222a99f5 - https://godotengine.org
OpenGL API 3.3.0 NVIDIA 536.40 - Compatibility - Using Device: NVIDIA Corporation - NVIDIA GeForce RTX 2060 with Max-Q Design
 
USER ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:972)
```

The crash happens every time when changing to a certain scene `with get_tree().change_scene_to_file()`. However, that scene I'm changing to, won't crash the application if it's set as the main one: the application will start normally. It's only when I try changing to the problematic scene, the crash happens.

When exporting for Windows with debug enabled, the issue doesn't reproduce, and there's nothing in the logs.

**Comment by YuriSizov** [comment]
Reopening, because https://github.com/godotengine/godot/pull/78987 is getting reverted.

**Comment by miv391** [comment]
I get this error occasionally just by starting Godot and loading a project. Probability is something like 10-20%. Sometimes a shiny version appears with two identical error messages.

System: v4.1.1.stable.official [bd6af8e0e], Windows 10

```
Godot Engine v4.1.1.stable.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
--- Debug adapter server started ---
--- GDScript language server started ---
  ./core/object/object.h:986 - Condition "slot >= slot_max" is true. Returning: nullptr
```

**Comment by SpaceAttorney** [comment]
Found out what was causing my error, looks unrelated to OP: #79371

**Comment by Daniel-Chin** [comment]
> I get this error occasionally just by starting Godot and loading a project. Probability is something like 10-20%. Sometimes a shiny version appears with two identical error messages.
> 
> I loaded the first Asset Library Project (NameCharacter), but this bug doesn't seem to happen when I load that project. My own project is larger and takes a few more seconds to load. So the size of the proejct _may_ have something to do with is.
> 
> System: v4.1.1.stable.official [[bd6af8e](https://github.com/godotengine/godot/commit/bd6af8e0ea69167dd0627f3bd54f9105bda0f8b5)], Windows 10
> 
> ```
> Godot Engine v4.1.1.stable.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
> --- Debug adapter server started ---
> --- GDScript language server started ---
>   ./core/object/object.h:986 - Condition "slot >= slot_max" is true. Returning: nullptr
> ```

I have the same issue. Around 10% of the times when opening a project with Godot v4.1.1.stable.mono.official.bd6af8e0e, 
```
ERROR: Condition "plugins_list.has(p_plugin)" is true.
   at: add_plugin (editor/editor_node.cpp:8195)
ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:986)
ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:986)
ERROR: Condition "plugins_list.has(p_plugin)" is true.
   at: add_plugin (editor/editor_node.cpp:8195)
```
Those errors are not fatal. I seem to use Godot just fine after those four errors.

**Comment by gynura** [comment]
I am having this same error when building to linux or windows, again if I export it with DEBUG it works fine, really nasty bug and no real fix: 

> Reproduces with Godot 4.0.3 (Steam). Exporting for Windows, release. Both when exporting on a Windows and a Mac OS machine. The full log when running my exported application and performing steps to reproduce is just this:
> 
> ```
> Godot Engine v4.0.3.stable.official.5222a99f5 - https://godotengine.org
> OpenGL API 3.3.0 NVIDIA 536.40 - Compatibility - Using Device: NVIDIA Corporation - NVIDIA GeForce RTX 2060 with Max-Q Design
>  
> USER ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
>    at: get_instance (./core/object/object.h:972)
> ```
> 
> The crash happens every time when changing to a certain scene `with get_tree().change_scene_to_file()`. However, that scene I'm changing to, won't crash the application if it's set as the main one: the application will start normally. It's only when I try changing to the problematic scene, the crash happens.
> 
> When exporting for Windows with debug enabled, the issue doesn't reproduce, and there's nothing in the logs.

**Comment by Koyper** [comment]
Also crashes release builds in MacOS (arm64) about once every six executions. About one in four executions report no errors, the rest report a series of `ERROR: Condition "slot >= slot_max" is true. Returning: nullptr`.

The crash occassionally reports something like this:

`(39992,0x209421280) malloc: Incorrect checksum for freed object 0x7fb8ce37e900: probably modified after being freed.
Corrupt value: 0xb00007008ce358b5
(39992,0x209421280) malloc: *** set a breakpoint in malloc_error_break to debug`

All debug builds run perfectly and report no issues with verbose output.

**Comment by homhomhomhomhom** [comment]
> I am having this same error when building to linux or windows, again if I export it with DEBUG it works fine, really nasty bug and no real fix:
> 
> > Reproduces with Godot 4.0.3 (Steam). Exporting for Windows, release. Both when exporting on a Windows and a Mac OS machine. The full log when running my exported application and performing steps to reproduce is just this:
> > ```
> > Godot Engine v4.0.3.stable.official.5222a99f5 - https://godotengine.org
> > OpenGL API 3.3.0 NVIDIA 536.40 - Compatibility - Using Device: NVIDIA Corporation - NVIDIA GeForce RTX 2060 with Max-Q Design
> >  
> > USER ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
> >    at: get_instance (./core/object/object.h:972)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The crash happens every time when changing to a certain scene `with get_tree().change_scene_to_file()`. However, that scene I'm changing to, won't crash the application if it's set as the main one: the application will start normally. It's only when I try changing to the problematic scene, the crash happens.
> > When exporting for Windows with debug enabled, the issue doesn't reproduce, and there's nothing in the logs.
> 
> Edit: I was able to fix this by changing the scenes from load() to preload() and changing the scene with this code:
> 
> ```
> get_tree().change_scene_to_packed(scene_to_load)
> queue_free()
> ```

didn't fix this error for me :( 
I have the exact same error every time a scene change happens, even with change_scene_to_packed.
I have tried preloading and using @export var = PackedScene
but this error "ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:986)" is still happening

**Comment by Dunkhan** [comment]
This error occurs for me when I unloaded and loaded a bunch of new nodes into the tree, there was never a scene change involved. Fiddling with the way things were loaded sometimes helped but it appears to be populating the tree that is the issue and not scene loading directly.

**Comment by Koyper** [comment]
I was finally able to reproduce this bug and prevent it from happening (so far). In my case it is caused by removing children from a parent, immediately followed by adding new children to the same parent.

Internally, the parent node appears very sensitive to being blocked by removing children while adding a new child is being attempted.

Only this combination of steps prevents the errors:

1. Children <em>must</em> be removed using `remove_child()`, followed by `queue_free()`. The error occurs either by omitting the `remove_child()` step, or using `free()` instead of `queue_free()`.
2. The new children must be then be added via a separate function called with `call_deferred("_add_new_children")`. Adding the children in a `for` loop using `call_deferred("add_children", new_child)` <em>does not</em> work.
3. Inside the `_add_new_children()` function, prefer to use `call_deferred("add_children", new_child), even though it so far seems to work without the deferred call. The bug is intermittent and why give it fuel?

It may also work to use a process, while or await to test that the children have been completely removed and freed prior to adding new children. If you need to handle a child after it's been added to the tree, you can use the `child_entered_tree` signal on the parent, which passes in the added child.

**Comment by Dunkhan** [comment]
Your reproduction matches my experience very closely. As far as the solution goes I found that using multiple defers did not always solve the problem though. This might be because I am using threads in c# so I was having difficulty establishing exactly when things happen and working out how to ensure that everything happens in the right order.

Hopefully this helps someone debug the internal functions though so we can make the behaviour more stable in the engine.

**Comment by Capital-EX** [comment]
> > so the main issue seems to me that this error message is entirely cryptic to users.
> 
> According to the code comments, "This should never happen unless RID is corrupted." So it cannot really be made less cryptic, since this seems to be a critical failure causing invalid data. Would be amazing to get a small, purpose-built MRP to find the culprit.

I'm not sure what happened between Godot 3 and Godot 4, but I have managed to make a MRP that seems to consistently triggers the bug when exported. @YuriSizov hopefully this can be of some use.

This code will trigger 2 error messages:
```gd
for i in range(100):
	var label = Label.new()
	label.text = "%d" % i
	$VBoxContainer.add_child(label)

for child in $VBoxContainer.get_children():
	child.free()
```

This code however will cause a spew of error message:
```gd
for i in range(100):
	var label = Label.new()
	label.text = "%d" % i
	$VBoxContainer.add_child(label)

for child in $VBoxContainer.get_children():
	child.free()

for i in range(100):
	var label = Label.new()
	label.text = "-%d" % i
	$VBoxContainer.add_child(label)
```

I have currently "fixed" this by purging `free` from every line of code in my project and replacing it with `queue_free`.

### MRP
[Freebug.zip](https://github.com/godotengine/godot/files/12877311/Freebug.zip)

**Comment by KccbzZ** [comment]
I have this same exact issue, the game crashes when I have a release build for Windows and never in debug or in the editor. It's really frustrating because I have turned my project that I worked on for about 2 years inside-out just because of trying to hunt down this error... It seems to appear during run time when I change a scene..

As mentioned before by other people here, using `free()` instead of `queue_free()` also seems to create this error. I also tried to use `call_deferred()` when changing scenes, and I think it does seem to minimize the error in a (semi-)test scene to only print the error once rather than 3-5 times at a time, and sometimes it doesn't print it at all. The game still crashes upon scene change but less frequently.

I should probably say I'm using Godot 4.1.1.stable, and here's the error because why not:

![image](https://github.com/godotengine/godot/assets/71567026/48171773-9827-4894-a37b-7260a60028ef)

I'd love to have a fix to this, I really don't want my project of 2 years to go to waste. And this error seems to appear for quite a few people so this does seem to need fixing.

**Comment by pshepjr** [comment]
I am getting this error too and sad to see that there is apparently no fix for it... :(

**Comment by KccbzZ** [comment]
> I am getting this error too and sad to see that there is apparently no fix for it... :(

@pshepjr In case your issue is also crashing upon scene change, see this comment from **gmjosack** https://github.com/godotengine/godot/issues/75422#issuecomment-1637120709, it seems to have fixed it for me!

In terms of the `"slot >= slot_max"` error, gmjosack uses `remove_child(child)` before doing `child.queue_free()`, so perhaps that could be of some assistance too.

I read that for Godot 4.2 they'll be focusing on stability so that's nice.

Hope that helps mate!

**Comment by dogboydog** [comment]
I had a crash following this error  in 4.2 dev 5, only in release builds. 

My code causing the crash was like this: 

```
        root.CallDeferred("add_child", sceneInstance);
        foreach (var node in nodesToFree)
        {
            if (!IsInstanceValid(node)){ continue;}
            node.CallDeferred("queue_free");
        }

```

Changing to this (calling QueueFree() immediately instead of deferring it) seems to avoid the crash on Windows and Linux 

```
       root.CallDeferred("add_child", sceneInstance);
        foreach (var node in nodesToFree)
        {
            if (!IsInstanceValid(node))
            {
                continue;
            }
            node.QueueFree();
        }
```

**Comment by akien-mga** [comment]
I can confirm the issue in 4.2-rc1 with @Capital-EX's MRP from https://github.com/godotengine/godot/issues/70910#issuecomment-1758935677

Actually with that project in 4.1.3-stable Linux release template I get an endless error spam of the `slot >= slot_max` error, while in 4.2-rc1 I get a short burst of spamming that error (~10 times) and then a different error and a segmentation fault:
```
ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:1033)
ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:1033)
ERROR: Condition "slot >= slot_max" is true. Returning: nullptr
   at: get_instance (./core/object/object.h:1033)
ERROR: Parameter "canvas_item" is null.
   at: canvas_item_clear (servers/rendering/renderer_canvas_cull.cpp:1558)
Segmentation fault (core dumped)
```

I'll try to compile a release template with debug symbols to get a stacktrace.
